CYF-ITP - South Africa | Simphiwe Mabuya | Structuring-and-Testing-Data | Sprint-3

Simphiwe-Mabuya opened the pull request:
<!--

You must title your PR like this:

COHORT_NAME | FIRST_NAME LAST_NAME | REPO_NAME | WEEK

For example,

NW4 | Carol Owen | HTML-CSS-Module | Week1

Complete the task list below this message.
If your PR is rejected, check the task list.

-->

## Learners, PR Template

Self checklist

- [ ] I have committed my files one by one, on purpose, and for a reason
- [ ] I have titled my PR with COHORT_NAME | FIRST_NAME LAST_NAME | REPO_NAME | WEEK 
- [ ] I have tested my changes
- [ ] My changes follow the [style guide](https://curriculum.codeyourfuture.io/guides/contributing/)
- [ ] My changes meet the [requirements](./README.md) of this task

## Changelist

Briefly explain your PR.

## Questions

Ask any questions you have for your reviewer.

cjyuan commented on Sprint-3/revise/implement/is-prime.test.js:
```diff
@@ -1,3 +1,63 @@
+
+});
+
+test("Checks if the number 13 is a prime number", () => {
```
Could also test some large prime and non-prime numbers.

cjyuan commented on Sprint-3/revise/implement/isPrime.js:
```diff
@@ -0,0 +1,16 @@
+        return `${num} is not prime`;
+    }
+    
+    for(let i = 2; i <= Math.sqrt(num); i++ ){
```
You can possibly improve the performance of the code in the following manners:
- Check if num is 2, and check only odd numbers >= 3 in the loop
- Avoid calling `Math.sqrt(num)` repeatedly by assigning the value of `Math.sqrt(num)` to a variable once, and then refer to the variable in the condition of the loop.
  - Note: The condition is checked at the start of every iteration.

cjyuan commented on Sprint-3/revise/implement/password-validator.test.js:
```diff
@@ -14,3 +14,30 @@ To be valid, a password must:
+
+test('Checks if password has at least 5 characters', () => {
+    const passwords = [];
+    const currentOutput = passwordValidator("B1@#", passwords);
```
`passwordValidator("B1@#", passwords)` can also return false because "B1@#" does not have a lowercase letter.

cjyuan commented on Sprint-3/revise/implement/password-validator.test.js:
```diff
@@ -14,3 +14,30 @@ To be valid, a password must:
+const passwordValidator = require("./passwordValidator");
+
+
+test('Checks if password is valid', () => {
```
This test does not check if the function can detect an invalid password that misses only one of the following conditions:
- Have at least one English uppercase letter (A-Z)
- Have at least one English lowercase letter (a-z)
- Have at least one number (0-9)
- Have at least one of the following non-alphanumeric symbols: ("!", "#", "$", "%", ".", "*", "&")

If you prepare tests to check if your function can detect each of the above cases separately, you may detect **a bug in your implementation**.

cjyuan commented on Sprint-3/revise/investigate/find.js:
```diff
@@ -20,6 +20,10 @@ console.log(find("code your future", "z"));
 // Pay particular attention to the following:
 
 // a) How the index variable updates during the call to find
+// 1. It updates as it iterates through the str when find() is called and will run / increment at least once while the condition is true or a character match is found.
```
The explanations given in 1, 2, 3, 4 are not very clear; some of them seems incorrect (to my understanding).

Can you try using ChatGPT to see how else you can explain the code?

cjyuan reviewed (commented):
I left my comments in the code.

Simphiwe-Mabuya commented:
Greetings CJ ..

I trust this email finds you well.

Thanks a million for your speedy feedback as always.

I’ll work on the items you’ve identified and get back to you as soon
as possible.

Enjoy the rest of your day.

Best regards
Simphiwe Mabuya

On Mon, 02 Dec 2024 at 20:29, CJ Yuan ***@***.***> wrote:

> ***@***.**** commented on this pull request.
>
> I left my comments in the code.
> ------------------------------
>
> In Sprint-3/implement/get-angle-type.test.js
> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1866285052>
> :
>
> > +test("Checks for Acute angle for when the angle is less than 90 degrees", () => {
> +    expect(currentOutput2).toEqual(targetOutput2);
> +});
> +
> +const currentOutput3 = getAngleType(145);
> +const targetOutput3 = "Obtuse angle";
> +
> +test("Checks for Obtuse angle for when the angle is greater than 90 and less than 180 degrees", () => {
> +    expect(currentOutput3).toEqual(targetOutput3);
> +});
> +
> +const currentOutput4 = getAngleType(180);
> +const targetOutput4 = "Straight angle";
> +
> +test("Checks for Straight angle for when the angle is exactly 180 degrees", () => {
> +    expect(currentOutput4).toEqual(targetOutput4);
>
> Could also write line 29 as
> expect(getAngleType(180)).toEqual("Straight angle");
> without first assigning the values to some variables first.
>
> Personally I find this style of writing makes the code less clutter.
> ------------------------------
>
> In Sprint-3/implement/get-angle-type.test.js
> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1866287310>
> :
>
> > +
> +test("Checks for Obtuse angle for when the angle is greater than 90 and less than 180 degrees", () => {
> +    expect(currentOutput3).toEqual(targetOutput3);
> +});
> +
> +const currentOutput4 = getAngleType(180);
> +const targetOutput4 = "Straight angle";
> +
> +test("Checks for Straight angle for when the angle is exactly 180 degrees", () => {
> +    expect(currentOutput4).toEqual(targetOutput4);
> +});
> +
> +const currentOutput5 = getAngleType(192);
> +const targetOutput5 = "Reflex angle";
> +
> +test("Checks for Reflex angle for when the angle is greater than 180 and less than 360 degrees", () => {
>
> Concise description is typically preferred.
>
> For example, you could describe this test case as:
> "Expect 'Reflect angle' to be returned when 180 < angle < 360", or even
> just
> "Expect 'Reflect angle' when 180 < angle < 360".
> ------------------------------
>
> In Sprint-3/implement/get-card-value.js
> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1866298176>
> :
>
> > +    }else if(cardRank === "J" || cardRank === "K" || cardRank === "Q" || cardRank === 10){
> +        return 10;
> +    }
> +
> +    const numericalValue = parseInt(cardRank, 10);
> +    if(numericalValue >= 2 && numericalValue <= 10){
> +        return numericalValue;
> +    }else {
> +        return 'Invalid card rank'
> +    }
> +}
> +module.exports = getCardValue;
> +
> +const assert = require("assert");
> +
> +function testGetCardValue(){
>
> What do you expect from the following function calls?
> Does your function returns the value you expected?
>
> getCardValue("AA♦");
> getCardValue("JK♦");
> getCardValue("0Q♦");
> getCardValue("02♦");
> getCardValue("2.1♦");
> getCardValue("10♦");
>
>
> ------------------------------
>
> In Sprint-3/implement/is-proper-fraction.js
> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1866311682>
> :
>
> > @@ -5,6 +5,20 @@
>  // Fractions: https://www.bbc.co.uk/bitesize/topics/zt9n6g8/articles/zjxpp4j
>  // Written here like this: 1/2 == Numerator/Denominator
>
> +function isProperFraction(num,den){
> +
> +    if(den <= 0){
>
> denominator can also be negative.
> ------------------------------
>
> In Sprint-3/implement/is-proper-fraction.test.js
> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1866320505>
> :
>
> > @@ -0,0 +1,33 @@
> +const isProperFraction = require("./is-proper-fraction");
> +
> +const currentOutput = isProperFraction(1, 2);
> +const targetOutput = true;
> +
> +test("Checks if the fraction is a Proper fraction (2/3) then return true", () => {
> +    expect(currentOutput).toEqual(targetOutput);
> +});
> +
> +const currentOutput2 = isProperFraction(-4, 7);
>
> isProperFraction(4, -7) should be considered as a proper fraction.
>
> The spec doesn't explicitly define what a "proper fraction" is, so it is
> up to you if you want to recognise isProperFraction(-4, -7) as a proper
> fraction.
> ------------------------------
>
> In Sprint-3/implement/is-valid-triangle.js
> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1866328934>
> :
>
> > @@ -11,6 +11,19 @@
>  // It's also true that b + c > a
>  // It's also true that a + c > b
>
> +function isValidTriangle(a,b,c){
> +
> +    if ( (a + b > c) && (a + c > b) && (b + c > a)){
> +        return true;
> +    }else if((a + b <= c) || (a + c <= b) || (b + c <= a)){
>
> According to De Morgan's Law, (a + b <= c) || (a + c <= b) || (b + c <= a)
> is the opposite of (a + b > c) && (a + c > b) && (b + c > a). So an else
> would be suffice.
>
> May I suggest looking up "Brief explanation of De Morgan's Law" via Google
> or ChatGPT?
> ------------------------------
>
> In Sprint-3/revise/implement/card-validator.test.js
> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1866337520>
> :
>
> > @@ -0,0 +1,17 @@
> +
> +const cardValidator = require("./cardValidator");
> +
> +
> +test("Checks if the number is 16 digits and all are numbers, at least two different digits are represented adn the sum of all digits is greater than 16", () => {
>
> When this test fails, how can you tell which of these three conditions the
> function failed to detect?
> ------------------------------
>
> In Sprint-3/revise/implement/is-prime.test.js
> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1866345389>
> :
>
> > +    const currentOutput = isPrime(7);
> +    const targetOutput = "7 is prime!";
> +
> +    expect(currentOutput).toBe(targetOutput);
> +
> +});
> +
> +test("Checks if the number 9 is a prime number", () => {
> +    const currentOutput = isPrime(9);
> +    const targetOutput = "9 is not prime";
> +
> +    expect(currentOutput).toBe(targetOutput);
> +
> +});
> +
> +test("Checks if the number 13 is a prime number", () => {
>
> Could also test some large prime and non-prime numbers.
> ------------------------------
>
> In Sprint-3/revise/implement/isPrime.js
> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1866346340>
> :
>
> > @@ -0,0 +1,16 @@
> +function isPrime(num){
> +    if(num <= 1){
> +        return `${num} is not prime`;
> +    }
> +
> +    for(let i = 2; i <= Math.sqrt(num); i++ ){
>
> You can possibly improve the performance of the code in the following
> manners:
>
>    - Check if num is 2, and check only odd numbers >= 3 in the loop
>    - Avoid calling Math.sqrt(num) repeatedly by assigning the value of
>    Math.sqrt(num) to a variable once, and then refer to the variable in
>    the condition of the loop.
>       - Note: The condition is checked at the start of every iteration.
>
> ------------------------------
>
> In Sprint-3/revise/implement/password-validator.test.js
> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1866360560>
> :
>
> > +    const targetOutput = true;
> +
> +    expect(currentOutput).toBe(targetOutput);
> +});
> +
> +test('Checks if password is already used', () => {
> +    const passwords = ["Si@#M3.pro!*z4"];
> +    const currentOutput = passwordValidator("Si@#M3.pro!*z4", passwords);
> +    const targetOutput = "Password is already used";
> +
> +    expect(currentOutput).toBe(targetOutput);
> +});
> +
> +test('Checks if password has at least 5 characters', () => {
> +    const passwords = [];
> +    const currentOutput = passwordValidator("B1@#", passwords);
>
> passwordValidator("B1@#", passwords) can also return false because "B1@#"
> does not have a lowercase letter.
> ------------------------------
>
> In Sprint-3/revise/implement/password-validator.test.js
> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1866366949>
> :
>
> > @@ -14,3 +14,30 @@ To be valid, a password must:
>
>  You must breakdown this problem in order to solve it. Find one test case first and get that working
>  */
> +
> +const passwordValidator = require("./passwordValidator");
> +
> +
> +test('Checks if password is valid', () => {
>
> This test does not check if the function can detect an invalid password
> that misses only one of the following conditions:
>
>    - Have at least one English uppercase letter (A-Z)
>    - Have at least one English lowercase letter (a-z)
>    - Have at least one number (0-9)
>    - Have at least one of the following non-alphanumeric symbols: ("!",
>    "#", "$", "%", ".", "*", "&")
>
> If you prepare tests to check if your function can detect each of the
> above cases separately, you may detect *a bug in your implementation*.
> ------------------------------
>
> In Sprint-3/revise/investigate/find.js
> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1866380642>
> :
>
> > @@ -20,6 +20,10 @@ console.log(find("code your future", "z"));
>  // Pay particular attention to the following:
>
>  // a) How the index variable updates during the call to find
> +// 1. It updates as it iterates through the str when find() is called and will run / increment at least once while the condition is true or a character match is found.
>
> The explanations given in 1, 2, 3, 4 are not very clear; some of them
> seems incorrect (to my understanding).
>
> Can you try using ChatGPT to see how else you can explain the code?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#pullrequestreview-2473398202>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BIXBQLKFHLKWU2BLOCPWJV32DSRHXAVCNFSM6AAAAABS3NMRN6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDINZTGM4TQMRQGI>
> .
> You are receiving this because you authored the thread.Message ID:
> <CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215/review/2473398202
> @github.com>
>

Simphiwe-Mabuya commented on Sprint-3/implement/get-angle-type.test.js:
```diff
@@ -0,0 +1,40 @@
+const targetOutput4 = "Straight angle";
+
+test("Checks for Straight angle for when the angle is exactly 180 degrees", () => {
+    expect(currentOutput4).toEqual(targetOutput4);
```
Thanks a million for this tip CJ ..I am certainly adopting it, it's pretty neat.

Simphiwe-Mabuya commented on Sprint-3/implement/get-card-value.js:
```diff
@@ -2,6 +2,47 @@
+
+const assert = require("assert");
+
+function testGetCardValue(){
```
I am still getting the values expected:

getCardValue("AA♦") returned value = 11
getCardValue("JK♦");  returned value = 10
getCardValue("0Q♦"); returned value = "Invalid card rank"
getCardValue("02♦"); returned value = "Invalid card rank"
getCardValue("2.1♦"); returned value = "Invalid card rank"
getCardValue("10♦"); returned value = "Invalid card rank"

Simphiwe-Mabuya commented on Sprint-3/implement/is-proper-fraction.js:
```diff
@@ -5,6 +5,20 @@
 
+function isProperFraction(num,den){
+    
+    if(den <= 0){
```
Fixed my code to handle negative denominator values

Simphiwe-Mabuya replied:
Fixed the code to handle negative denominator.

Simphiwe-Mabuya commented on Sprint-3/implement/is-proper-fraction.test.js:
```diff
@@ -0,0 +1,33 @@
+    expect(currentOutput).toEqual(targetOutput);
+});
+
+const currentOutput2 = isProperFraction(-4, 7);
```
I was also able to have my code handle isProperFraction(4,-7) or isProperFraction(-4,-7) as proper fractions. Thanks for the tip CJ, I wasn't actually aware that I could have my code handle negative denominator values this way.

Simphiwe-Mabuya commented on Sprint-3/implement/is-valid-triangle.js:
```diff
@@ -11,6 +11,19 @@
+
+    if ( (a + b > c) && (a + c > b) && (b + c > a)){
+        return true;
+    }else if((a + b <= c) || (a + c <= b) || (b + c <= a)){
```
Thanks yet again CJ for making me aware of the De Morgan's Law. I now have an understanding of what it means, thus I refactored my code accordingly without the opposite comparison.

cjyuan commented:
Can you reply my comments in the "conversation" that appear in the changed code directly, so that I know what changes you have made (without going through all the files)? The bug I mentioned in one of my comments is still in your code.

Simphiwe-Mabuya commented:
Greetings Mr CJ...I have been doing that actually, it's just that the
session I was replying to on GitHub refreshed and I lost it.

I now have the choice to reply from this email thread I guess.
I replied to our "conversation" up to :
Sprint-3/implement/is-valid-triangle.js
<https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1866328934>
from our "conversation" on GitHub.

In Sprint-3/revise/implement/card-validator.test.js
<https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1866337520>:

I have refactored my code to clean and wrote tests to check for: "Check if
the number has 16 digits", "Checks if the sum of all numbers is greater
than 16" and "Checks if at least two different digits are represented".

In Sprint-3/revise/implement/is-prime.test.js
<https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1866345389>:

Refactored code to look clean and tested for large prime and non-prime
numbers.
I also tried to improve the performance of the code by following your
suggestions:
-  - Check if num is 2, and check only odd numbers >= 3 in the loop
-  - Avoid calling Math.sqrt(num) repeatedly by assigning the value of
Math.sqrt(num) to a variable once, and then refer to the variable in the
condition of the loop.

In Sprint-3/revise/implement/password-validator.test.js
<https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1866360560>:

I've added a lower case letter here to check if the test will return false
and it does. I am not sure if I followed you well on this point ..? I
thought this test was to check if the password has at least five
characters.

I've also refactored the test to look clean plus have the test to check if
the function can detect an invalid password that misses only one of the
following conditions :

   - Have at least one English uppercase letter (A-Z)
   - Have at least one English lowercase letter (a-z)
   - Have at least one number (0-9)
   - Have at least one of the following non-alphanumeric symbols: ("!",
   "#", "$", "%", ".", "*", "&")

In Sprint-3/revise/investigate/find.js
<https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1866380642>:


I also updated my answers after I followed your advice by using ChatGpt.
I believe my answers should now be correct.

I trust that I've managed to answer and rectify all the items you have
identified.
Looking forward to your constructive feedback again.

Enjoy the rest of the evening.

Best regards
Simphiwe Mabuya




<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virus-free.www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

On Tue, 3 Dec 2024 at 20:15, CJ Yuan ***@***.***> wrote:

> Can you reply my comments in the "conversation" that appear in the changed
> code directly, so that I know what changes you have made (without going
> through all the files)? The bug I mentioned in one of my comments is still
> in your code.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#issuecomment-2515270790>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BIXBQLNLKENJN52Y66OC6RT2DXYLBAVCNFSM6AAAAABS3NMRN6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKMJVGI3TANZZGA>
> .
> You are receiving this because you authored the thread.Message ID:
> <CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215/c2515270790@
> github.com>
>

cjyuan commented on Sprint-3/revise/implement/card-validator.test.js:
```diff
@@ -0,0 +1,23 @@
+
+
+test("Checks if the number has 16 digits.", () => {
+    expect(cardValidator("1243578569873476")).toEqual(true)
```
The function can successfully check if there are 16 digits in the card number does not mean that the function can return false when the card number has 15 or 20 digits.

cjyuan commented on Sprint-3/revise/implement/isPrime.js:
```diff
@@ -0,0 +1,22 @@
+function isPrime(num){
```
Why not design the function to return `true` or `false` to indicate if the given number is a prime?

cjyuan commented on Sprint-3/revise/implement/password-validator.test.js:
```diff
@@ -14,3 +14,36 @@ To be valid, a password must:
+
+test('Checks if password has at least one English uppercase letter (A-Z)', () => {
+    const passwords =[];
+    expect(passwordValidator("Bad@11.#YoZ%", passwords)).toBe(true);
```
When given a valid password, the function can correctly return true. (This is what your test represents)

When given an invalid password, can the function correctly return false? (This is what I would like you to test)
For examples, a password missing only a digit, a password missing only an alphanumeric character, etc.

If you prepare the tests correctly, you may discover the bug in your code.

cjyuan commented on Sprint-3/revise/implement/passwordValidator.js:
```diff
@@ -0,0 +1,28 @@
+     }
+
+     if(passwords.includes(password)){
+        return "Password is already used"
```
Better to return `false` to make the return value consistent.

Simphiwe-Mabuya commented:
Good morning CJ

Thanks a lot for your feedback as always..it looks like I’ve got some work
to do.
Let me get on it and get back to you as soon as I can.

Best regards
Simphiwe Mabuya

On Wed, 04 Dec 2024 at 03:21, CJ Yuan ***@***.***> wrote:

> ***@***.**** commented on this pull request.
>
> I let you new comments.
> ------------------------------
>
> In Sprint-3/implement/get-angle-type.js
> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1868525108>
> :
>
> > @@ -6,6 +6,21 @@
>  // When the function getAngleType is called with this angle,
>  // Then it should:
>
> +function getAngleType(degrees) {
> +    if (degrees === 90) return "Right angle";
> +    if (degrees < 90) return "Acute angle";
>
> What do you expect getAngleType(-5) to return?
> ------------------------------
>
> In Sprint-3/implement/get-angle-type.test.js
> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1868525888>
> :
>
> > +test("Checks for Acute angle, when the angle < 90 degrees", () => {
> +    expect(getAngleType(45)).toEqual("Acute angle");
> +});
> +
> +
> +test("Checks for Obtuse angle, when the angle > 90 and angle < 180 degrees", () => {
> +    expect(getAngleType(145)).toEqual("Obtuse angle");
> +});
> +
> +
> +test("Checks for Straight angle, when the angle is exactly 180 degrees", () => {
> +    expect(getAngleType(180)).toEqual("Straight angle");
> +});
> +
> +
> +test("Expect 'Reflect angle' when 180 < angle < 360", () => {
>
> I was just making a suggestion so that you can try to keep the test
> description concise in the future.
> ------------------------------
>
> In Sprint-3/implement/get-card-value.js
> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1868545061>
> :
>
> > @@ -2,6 +2,37 @@
>
>  // You will need to implement a function getCardValue
>
> +function getCardValue(cardRank){
>
> getCardValue("10000♣") would still break your implementation. (sorry)
>
> The card value should always have ONE suite character as the LAST
> character. So
> getCardValue("10") should return 'Invalid card rank'.
>
> Since there are exactly 13 valid ranks but countless possible invalid
> ranks,
> it would be easier to consider what value constitutes a valid rank
> instead of considering what values are invalid.
> After you have checked all valid cases, everything else would be invalid.
> ------------------------------
>
> In Sprint-3/implement/get-card-value.js
> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1868546010>
> :
>
> > @@ -2,6 +2,37 @@
>
>  // You will need to implement a function getCardValue
>
> +function getCardValue(cardRank){
> +
> +    if(cardRank === "10"){
>
> If you can remove the suite character from cardRank first, this is a good
> way to guarantee the return value is correct.
> ------------------------------
>
> In Sprint-3/implement/get-card-value.js
> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1868548445>
> :
>
> > @@ -2,6 +2,37 @@
>
>  // You will need to implement a function getCardValue
>
> +function getCardValue(cardRank){
> +
> +    if(cardRank === "10"){
> +        return 10;
> +    }
> +
> +    if(/^0\d/.test(cardRank) || /\.\d/.test(cardRank)){
>
> You can skip checking for possible invalid cases.
> ------------------------------
>
> In Sprint-3/implement/get-card-value.js
> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1868550610>
> :
>
> > +    }
> +
> +    if(/^0\d/.test(cardRank) || /\.\d/.test(cardRank)){
> +        return "Invalid card rank";
> +    }
> +
> +    cardRank = cardRank.slice(0, 2) === "10" ? "10" : cardRank.slice(0, 1);
> +
> +    if(cardRank ==="A"){
> +        return 11;
> +    }else if(cardRank === "J" || cardRank === "K" || cardRank === "Q"){
> +        return 10;
> +    }
> +
> +    const numericalValue = parseInt(cardRank, 10);
> +    if(numericalValue >= 2 && numericalValue <= 10){
>
> You have already checked for "10" at line 7. So the remaining values to be
> checked are "2", "3", ..., "9". What do these strings have in common? What
> condition can you add so that ranks like "2.1" would not pass the check.
> ------------------------------
>
> In Sprint-3/implement/get-card-value.js
> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1868552528>
> :
>
> > @@ -2,6 +2,37 @@
>
>  // You will need to implement a function getCardValue
>
> +function getCardValue(cardRank){
> +
> +    if(cardRank === "10"){
> +        return 10;
> +    }
> +
> +    if(/^0\d/.test(cardRank) || /\.\d/.test(cardRank)){
> +        return "Invalid card rank";
> +    }
> +
> +    cardRank = cardRank.slice(0, 2) === "10" ? "10" : cardRank.slice(0, 1);
>
> This part should be replaced by a statement that removes the suite
> character from the string, and should be perform at the beginning of the
> function.
> ------------------------------
>
> In Sprint-3/revise/implement/card-validator.test.js
> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1868555940>
> :
>
> > @@ -0,0 +1,23 @@
> +
> +const cardValidator = require("./cardValidator");
> +
> +
> +test("Checks if the number has 16 digits.", () => {
> +    expect(cardValidator("1243578569873476")).toEqual(true)
>
> The function can successfully check if there are 16 digits in the card
> number does not mean that the function can return false when the card
> number has 15 or 20 digits.
> ------------------------------
>
> In Sprint-3/revise/implement/isPrime.js
> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1868562023>
> :
>
> > @@ -0,0 +1,22 @@
> +function isPrime(num){
>
> Why not design the function to return true or false to indicate if the
> given number is a prime?
> ------------------------------
>
> In Sprint-3/revise/implement/password-validator.test.js
> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1868566402>
> :
>
> > @@ -14,3 +14,36 @@ To be valid, a password must:
>
>  You must breakdown this problem in order to solve it. Find one test case first and get that working
>  */
> +
> +const passwordValidator = require("./passwordValidator");
> +
> +
> +test('Checks if password has at least one English uppercase letter (A-Z)', () => {
> +    const passwords =[];
> +    ***@***.***#YoZ%", passwords)).toBe(true);
>
> When given a valid password, the function can correctly return true. (This
> is what your test represents)
>
> When given an invalid password, can the function correctly return false?
> (This is what I would like you to test)
> For examples, a password missing only a digit, a password missing only an
> alphanumeric character, etc.
>
> If you prepare the tests correctly, you may discover the bug in your code.
> ------------------------------
>
> In Sprint-3/revise/implement/passwordValidator.js
> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1868567013>
> :
>
> > @@ -0,0 +1,28 @@
> +function passwordValidator(password, passwords){
> +
> +    if(password.length < 5){
> +        return false;
> +     }
> +
> +     if(passwords.includes(password)){
> +        return "Password is already used"
>
> Better to return false to make the return value consistent.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#pullrequestreview-2477000444>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BIXBQLNJCVDSVSL3OJKOWND2DZKIHAVCNFSM6AAAAABS3NMRN6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDINZXGAYDANBUGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> <CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215/review/2477000444
> @github.com>
>

Simphiwe-Mabuya commented on Sprint-3/implement/get-angle-type.test.js:
```diff
@@ -0,0 +1,30 @@
+});
+
+
+test("Expect 'Reflect angle' when 180 < angle < 360", () => {
```
Thanks for the tip CJ ..I thought I should used it.

Simphiwe-Mabuya commented:
Greetings CJ again

I hope this finds you well this afternoon.

Here are my solutions as per your directions/feedback:

In Sprint-3/implement/get-angle-type.js
<https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1868525108>:


I expect getAngleType(-5) to return 'Acute Angle' because  -5 is less than
90 (-5 < 90)

In Sprint-3/implement/get-card-value.js
<https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1868545061>:


- I believe I've managed to check all valid cases are working and have
checked everything else
would be invalid..eg "10", "2.1", "Z", "10000 ♣ " etc
- I also made sure to remove the suite character from cardRank first right
from the start.
- Also I made sure to skip checking for possible invalid cases with regex.
-  I also made sure that my code checks for "2", "3", ..., "9" ...and the
thing that's common with these strings is that they are all. I used
'numericalValue.toString() === cardRank' to make sure "2.1" does not pass.

In Sprint-3/revise/implement/card-validator.test.js
<https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1868555940>:


- I've updated my test to check for the card number when it's less than 16
or greater (15 / 20 ) and returns false if so.

In Sprint-3/revise/implement/isPrime.js
<https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1868562023>:


I've updated the function to return true or false to indicate if the given
number is a prime or not as advised.

In Sprint-3/revise/implement/password-validator.test.js
<https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1868566402>:


Also  updated my password-validator test to check for when the password is
missing only a digit or only an alphanumeric character and to return false
if so.

In Sprint-3/revise/implement/passwordValidator.js
<https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1868567013>:


and lastly I've updated the function to return false  istead of "Password
already used"  to make the return value consistent.


Thanks a million yet again CJ for your constructive feedback, looking
forward to hearing from you again.
Please do enjoy the rest of the day.

Best regards
Simphiwe Mabuya






On Wed, 4 Dec 2024 at 09:48, Simphiwe Mabuya ***@***.***> wrote:

> Good morning CJ
>
> Thanks a lot for your feedback as always..it looks like I’ve got some work
> to do.
> Let me get on it and get back to you as soon as I can.
>
> Best regards
> Simphiwe Mabuya
>
> On Wed, 04 Dec 2024 at 03:21, CJ Yuan ***@***.***> wrote:
>
>> ***@***.**** commented on this pull request.
>>
>> I let you new comments.
>> ------------------------------
>>
>> In Sprint-3/implement/get-angle-type.js
>> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1868525108>
>> :
>>
>> > @@ -6,6 +6,21 @@
>>  // When the function getAngleType is called with this angle,
>>  // Then it should:
>>
>> +function getAngleType(degrees) {
>> +    if (degrees === 90) return "Right angle";
>> +    if (degrees < 90) return "Acute angle";
>>
>> What do you expect getAngleType(-5) to return?
>> ------------------------------
>>
>> In Sprint-3/implement/get-angle-type.test.js
>> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1868525888>
>> :
>>
>> > +test("Checks for Acute angle, when the angle < 90 degrees", () => {
>> +    expect(getAngleType(45)).toEqual("Acute angle");
>> +});
>> +
>> +
>> +test("Checks for Obtuse angle, when the angle > 90 and angle < 180 degrees", () => {
>> +    expect(getAngleType(145)).toEqual("Obtuse angle");
>> +});
>> +
>> +
>> +test("Checks for Straight angle, when the angle is exactly 180 degrees", () => {
>> +    expect(getAngleType(180)).toEqual("Straight angle");
>> +});
>> +
>> +
>> +test("Expect 'Reflect angle' when 180 < angle < 360", () => {
>>
>> I was just making a suggestion so that you can try to keep the test
>> description concise in the future.
>> ------------------------------
>>
>> In Sprint-3/implement/get-card-value.js
>> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1868545061>
>> :
>>
>> > @@ -2,6 +2,37 @@
>>
>>  // You will need to implement a function getCardValue
>>
>> +function getCardValue(cardRank){
>>
>> getCardValue("10000♣") would still break your implementation. (sorry)
>>
>> The card value should always have ONE suite character as the LAST
>> character. So
>> getCardValue("10") should return 'Invalid card rank'.
>>
>> Since there are exactly 13 valid ranks but countless possible invalid
>> ranks,
>> it would be easier to consider what value constitutes a valid rank
>> instead of considering what values are invalid.
>> After you have checked all valid cases, everything else would be invalid.
>> ------------------------------
>>
>> In Sprint-3/implement/get-card-value.js
>> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1868546010>
>> :
>>
>> > @@ -2,6 +2,37 @@
>>
>>  // You will need to implement a function getCardValue
>>
>> +function getCardValue(cardRank){
>> +
>> +    if(cardRank === "10"){
>>
>> If you can remove the suite character from cardRank first, this is a
>> good way to guarantee the return value is correct.
>> ------------------------------
>>
>> In Sprint-3/implement/get-card-value.js
>> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1868548445>
>> :
>>
>> > @@ -2,6 +2,37 @@
>>
>>  // You will need to implement a function getCardValue
>>
>> +function getCardValue(cardRank){
>> +
>> +    if(cardRank === "10"){
>> +        return 10;
>> +    }
>> +
>> +    if(/^0\d/.test(cardRank) || /\.\d/.test(cardRank)){
>>
>> You can skip checking for possible invalid cases.
>> ------------------------------
>>
>> In Sprint-3/implement/get-card-value.js
>> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1868550610>
>> :
>>
>> > +    }
>> +
>> +    if(/^0\d/.test(cardRank) || /\.\d/.test(cardRank)){
>> +        return "Invalid card rank";
>> +    }
>> +
>> +    cardRank = cardRank.slice(0, 2) === "10" ? "10" : cardRank.slice(0, 1);
>> +
>> +    if(cardRank ==="A"){
>> +        return 11;
>> +    }else if(cardRank === "J" || cardRank === "K" || cardRank === "Q"){
>> +        return 10;
>> +    }
>> +
>> +    const numericalValue = parseInt(cardRank, 10);
>> +    if(numericalValue >= 2 && numericalValue <= 10){
>>
>> You have already checked for "10" at line 7. So the remaining values to
>> be checked are "2", "3", ..., "9". What do these strings have in common?
>> What condition can you add so that ranks like "2.1" would not pass the
>> check.
>> ------------------------------
>>
>> In Sprint-3/implement/get-card-value.js
>> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1868552528>
>> :
>>
>> > @@ -2,6 +2,37 @@
>>
>>  // You will need to implement a function getCardValue
>>
>> +function getCardValue(cardRank){
>> +
>> +    if(cardRank === "10"){
>> +        return 10;
>> +    }
>> +
>> +    if(/^0\d/.test(cardRank) || /\.\d/.test(cardRank)){
>> +        return "Invalid card rank";
>> +    }
>> +
>> +    cardRank = cardRank.slice(0, 2) === "10" ? "10" : cardRank.slice(0, 1);
>>
>> This part should be replaced by a statement that removes the suite
>> character from the string, and should be perform at the beginning of the
>> function.
>> ------------------------------
>>
>> In Sprint-3/revise/implement/card-validator.test.js
>> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1868555940>
>> :
>>
>> > @@ -0,0 +1,23 @@
>> +
>> +const cardValidator = require("./cardValidator");
>> +
>> +
>> +test("Checks if the number has 16 digits.", () => {
>> +    expect(cardValidator("1243578569873476")).toEqual(true)
>>
>> The function can successfully check if there are 16 digits in the card
>> number does not mean that the function can return false when the card
>> number has 15 or 20 digits.
>> ------------------------------
>>
>> In Sprint-3/revise/implement/isPrime.js
>> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1868562023>
>> :
>>
>> > @@ -0,0 +1,22 @@
>> +function isPrime(num){
>>
>> Why not design the function to return true or false to indicate if the
>> given number is a prime?
>> ------------------------------
>>
>> In Sprint-3/revise/implement/password-validator.test.js
>> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1868566402>
>> :
>>
>> > @@ -14,3 +14,36 @@ To be valid, a password must:
>>
>>  You must breakdown this problem in order to solve it. Find one test case first and get that working
>>  */
>> +
>> +const passwordValidator = require("./passwordValidator");
>> +
>> +
>> +test('Checks if password has at least one English uppercase letter (A-Z)', () => {
>> +    const passwords =[];
>> +    ***@***.***#YoZ%", passwords)).toBe(true);
>>
>> When given a valid password, the function can correctly return true.
>> (This is what your test represents)
>>
>> When given an invalid password, can the function correctly return false?
>> (This is what I would like you to test)
>> For examples, a password missing only a digit, a password missing only an
>> alphanumeric character, etc.
>>
>> If you prepare the tests correctly, you may discover the bug in your code.
>> ------------------------------
>>
>> In Sprint-3/revise/implement/passwordValidator.js
>> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#discussion_r1868567013>
>> :
>>
>> > @@ -0,0 +1,28 @@
>> +function passwordValidator(password, passwords){
>> +
>> +    if(password.length < 5){
>> +        return false;
>> +     }
>> +
>> +     if(passwords.includes(password)){
>> +        return "Password is already used"
>>
>> Better to return false to make the return value consistent.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#pullrequestreview-2477000444>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/BIXBQLNJCVDSVSL3OJKOWND2DZKIHAVCNFSM6AAAAABS3NMRN6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDINZXGAYDANBUGQ>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> <CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215/review/2477000444
>> @github.com>
>>
>

cjyuan commented:
The tests in `Sprint-3/revise/implement/password-validator.test.js` are still not comprehensive enough to detect the bug in your function. 

For a certain kind of invalid passwords, your current function is returning `true` instead of `false`. If you cannot figure out which test cases you missed or find the bug in your code, you can test different kinds of invalid passwords to see which kind of invalid passwords your function cannot detect.

Simphiwe-Mabuya commented:
Alright 😁 ..this is going to be an interesting yet challenging one.

Thanks a lot again CJ for such s speedy response 🙏🏾🙏🏾

Let me get on it and get back to you as soon as possible.

On Wed, 04 Dec 2024 at 19:15, CJ Yuan ***@***.***> wrote:

> The tests in Sprint-3/revise/implement/password-validator.test.js are
> still not comprehensive enough to detect the bug in your function.
>
> For a certain kind of invalid passwords, your current function is
> returning true instead of false. If you cannot figure out which test
> cases you missed or find the bug in your code, you can test different kinds
> of invalid passwords to see which kind of invalid passwords your function
> cannot detect.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#issuecomment-2518066266>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BIXBQLORVPT4A3OAXEGYAZD2D42C3AVCNFSM6AAAAABS3NMRN6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKMJYGA3DMMRWGY>
> .
> You are receiving this because you authored the thread.Message ID:
> <CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215/c2518066266@
> github.com>
>

Simphiwe-Mabuya commented:
Evening CJ ..I trust that this will find you well.

I think I've found the bug in my code..the invalid password was a password
with only one uppercase / lowercase letter that returned true instead of
false. Thus I fixed my regex expression to check if the password has at
least one uppercase and one lowercase letter.

Looking forward to hearing your thoughts regarding my attempted solution.

Kind regards
Simphiwe Mabuya

On Wed, 4 Dec 2024 at 20:13, Simphiwe Mabuya ***@***.***> wrote:

> Alright 😁 ..this is going to be an interesting yet challenging one.
>
> Thanks a lot again CJ for such s speedy response 🙏🏾🙏🏾
>
> Let me get on it and get back to you as soon as possible.
>
> On Wed, 04 Dec 2024 at 19:15, CJ Yuan ***@***.***> wrote:
>
>> The tests in Sprint-3/revise/implement/password-validator.test.js are
>> still not comprehensive enough to detect the bug in your function.
>>
>> For a certain kind of invalid passwords, your current function is
>> returning true instead of false. If you cannot figure out which test
>> cases you missed or find the bug in your code, you can test different kinds
>> of invalid passwords to see which kind of invalid passwords your function
>> cannot detect.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#issuecomment-2518066266>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/BIXBQLORVPT4A3OAXEGYAZD2D42C3AVCNFSM6AAAAABS3NMRN6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKMJYGA3DMMRWGY>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> <CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215/c2518066266@
>> github.com>
>>
>

Simphiwe-Mabuya commented:
Greetings CJ ..

I am sure this will find you well.
I just wanted to confirm if my last answer was the final issue to resolve
and if it’s now safe to mark my PR complete?

Thanks a million as always and I’m looking forward to your feedback in the
near future.

Kind regards
Simphiwe Mabuya

On Wed, 04 Dec 2024 at 23:05, Simphiwe Mabuya ***@***.***> wrote:

> Evening CJ ..I trust that this will find you well.
>
> I think I've found the bug in my code..the invalid password was a password
> with only one uppercase / lowercase letter that returned true instead of
> false. Thus I fixed my regex expression to check if the password has at
> least one uppercase and one lowercase letter.
>
> Looking forward to hearing your thoughts regarding my attempted solution.
>
> Kind regards
> Simphiwe Mabuya
>
> On Wed, 4 Dec 2024 at 20:13, Simphiwe Mabuya ***@***.***> wrote:
>
>> Alright 😁 ..this is going to be an interesting yet challenging one.
>>
>> Thanks a lot again CJ for such s speedy response 🙏🏾🙏🏾
>>
>> Let me get on it and get back to you as soon as possible.
>>
>> On Wed, 04 Dec 2024 at 19:15, CJ Yuan ***@***.***> wrote:
>>
>>> The tests in Sprint-3/revise/implement/password-validator.test.js are
>>> still not comprehensive enough to detect the bug in your function.
>>>
>>> For a certain kind of invalid passwords, your current function is
>>> returning true instead of false. If you cannot figure out which test
>>> cases you missed or find the bug in your code, you can test different kinds
>>> of invalid passwords to see which kind of invalid passwords your function
>>> cannot detect.
>>>
>>> —
>>> Reply to this email directly, view it on GitHub
>>> <https://github.com/CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215#issuecomment-2518066266>,
>>> or unsubscribe
>>> <https://github.com/notifications/unsubscribe-auth/BIXBQLORVPT4A3OAXEGYAZD2D42C3AVCNFSM6AAAAABS3NMRN6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKMJYGA3DMMRWGY>
>>> .
>>> You are receiving this because you authored the thread.Message ID:
>>> <CodeYourFuture/Module-Structuring-and-Testing-Data/pull/215/c2518066266
>>> @github.com>
>>>
>>

cjyuan commented:
Yes. That's my confirmation.
